flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/TaskExecutorPartitionTrackerImpl.java

```diff
@@ -24,12 +24,18 @@
 import org.apache.flink.util.Preconditions;
 
 import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 /**
  * Utility for tracking partitions and issuing release calls to task executors and shuffle masters.
  */
 public class TaskExecutorPartitionTrackerImpl extends AbstractPartitionTracker<JobID, TaskExecutorPartitionInfo> implements TaskExecutorPartitionTracker {
 
+	private final Map<DataSetMetaInfo, Set<ResultPartitionID>> clusterPartitions = new HashMap<>();
```

Comment:
Where do we need this indexing? for the report to RM? Is `IntermediateDataSetID` not enough for this indexing?
which other info will be added to `DataSetMetaInfo` and how it differs from `TaskExecutorPartitionInfo`?
Maybe, some comments could be added.

Reply:
yeeeah we can probably reuse the `TaskExecutorPartitionInfo`.

I'll have to change the indexing a bit since it currently doesn't make sense; we need a mapping from DataSetID -> ResultPartitionIDs to facilitate the release, but in the current form this isn't possible.
The metadata can be either be kept in a separate map, or the value of the map is a Tuple2 containing both the meta info and ResultPartitionIDs.

Reply:
Actually, we can keep the indexing as is so long as we only use the dataset in equals/hashCode. This would simplify things a bit.

Reply:
ok, got it, you are talking about next step where RM externally release a cluster partition. I guess it will be one more method.